js/background.js

```diff
@@ -41,61 +38,482 @@ const LocalStorage = (function(){
+        if (!timestamp) return true;
+        if (typeof ttl != 'number' || ttl < 0) ttl = 0;
+        return timestamp + ttl <= this.timestamp();
+    };
+
+    self.get = function(key, ttl, defaultValue) {
+        if (!ttl) return defaultValue;
+        let item = localStorage.getItem('cache_' + key);
+        if (!item) return defaultValue;
+        try {
+            item = JSON.parse(item);
+        } catch (err) {
+            return defaultValue;
+        }
+        if (!item.timestamp || self.isExpired(item.timestamp, ttl)) return defaultValue;
+        return item.data;
+    };
+
+    self.set = function(key, value) {
+        localStorage.setItem('cache_' + key, JSON.stringify({ 'data': value, 'timestamp': self.timestamp(), }));
+    };
+
+    self.remove = function(key) {
+        localStorage.removeItem('cache_' + key);
+    };
+
+    self.keys = function() {
+        return LocalStorage.keys()
+            .filter(k => k.startsWith('cache_'))
+            .map(k => k.substring(6)); // "cache_".length == 6
+    };
+
+    self.clear = function() {
+        let keys = self.keys();
+        for (let key of keys) {
+            self.remove(key);
+        }
+    };
+
+    Object.freeze(self);
+    return self;    
+})();
+
+
+class Api {
+    constructor(origin, params={}) {
+        if (!origin) throw `Constructor requires an Origin`;
+        this.origin = origin;
+        this.params = params;
+    }
+    // withResponse? use a boolean to include Response object in result?
+    fetch(endpoint, query={}, params={}) {
+        let url = new URL(endpoint, this.origin);
+        for (let [k, v] of Object.entries(query)) {
+            url.searchParams.append(k, v);
+        }
+        params = Object.assign({}, this.params, params);
+        return fetch(url, params);
+    }
+    getEndpoint(endpoint, query) {
+        if (!endpoint.endsWith('/'))
+            endpoint += '/';
+        return this.fetch(endpoint, query, { 'method': 'GET', }).then(response => response.json());
+    }    
+    getPage(endpoint, query) {
+        return this.fetch(endpoint, query, { 'method': 'GET', }).then(response => response.text());
+    }
+}
+
+
+const AugmentedSteamApi = (function() {
+    let self = new Api(Config.ApiServerHost);
+
+    let progressingRequests = new Map();
+    self._getEndpoint = self.getEndpoint;
+    self.getEndpoint = function(endpoint, query) { // withResponse? boolean that includes Response object in result?
+        return self._getEndpoint(endpoint, query)
+            .then(function(json) {
+                if (!json.result || json.result !== "success")
+                    throw `Could not retrieve '${endpoint}'`;
+                delete json.result;
+                return json; // 'response': response, 
+            })
+        ;
+    };
+
     /**
-     * Requires user to be signed in, can we validate this from background?
+     * self.prices = async function({ 'params': params }) {
```

Comment:
Isn't this already substituted by the factory function?

Reply:
Err, yes. The comment was a note-to-self of what `AugmentedSteamApi.endpointFactory()` was supposed to return. It also helped as a guide as to "can I replace this function, or is it doing more than the guide?".

Reply:
Fixed in c173602